deepsubtree.go

```diff
@@ -1,31 +1,48 @@
 package smt
 
-import(
-    "hash"
+import (
+	"hash"
 )
 
+// BadProofError is returned when an invalid Merkle proof is supplied.
+type BadProofError struct{}
+
+func (e *BadProofError) Error() string {
+	return "bad proof"
+}
```

Comment:
Similar to the other place, this is equivalent to a sentinel error:
```suggestion
// BadProofError is returned when an invalid Merkle proof is supplied.
var ErrBadProof = errors.New("bad proof")
```

Also, it is more common / go-idiomatic to have a `Err` prefix.

Reply:
Isn't it better to have a special type for the error, so that users of the API can differentiate between error types?

Reply:
In go it is very common to use sentinel values for cases like this (this custom error is just returns a string with no further fields or context). See for instance the io package: https://golang.org/pkg/io/#pkg-variables 

The custom error doesn't provide any additional value compared to a sentinel error (not to say that in other cases it could make more sense to use a custom error).